File: scripts/KeyCodes.cs
using Godot;
using System;
using System.Collections.Generic;
using Newtonsoft.Json;
using System.Linq;
using System.Text.RegularExpressions;

namespace Peg
{
    public class KeyCodes
    {
        static KeyCodes instance;
        public Dictionary<string, KeyCode> Basic;
        public Dictionary<string, KeyCode> LessUsed;
        public Dictionary<string, KeyCode> Bluetooth;
        public Dictionary<string, KeyCode> InternalCodes;
        public Dictionary<string, KeyCode> International;
        public Dictionary<string, KeyCode> Modifiers;
        public Dictionary<string, KeyCode> Shifted;
        public Dictionary<string, KeyCode> Layers;
        public Dictionary<string, KeyCode> Led;
        public Dictionary<string, KeyCode> CustomCodes;


        protected KeyCodes(){
            Godot.File file = new Godot.File();
            file.Open("res://jsonKeycodes/basic-keycodes.json", Godot.File.ModeFlags.Read);
            string basictext = file.GetAsText();
            file.Close();
            file.Open("res://jsonKeycodes/lessused-keycodes.json", Godot.File.ModeFlags.Read);
            string lessUsedtext = file.GetAsText();
            file.Close();
            file.Open("res://jsonKeycodes/bluetooth-keycodes.json", Godot.File.ModeFlags.Read);
            string bluetoothtext = file.GetAsText();
            file.Close();
            file.Open("res://jsonKeycodes/internal-keycodes.json", Godot.File.ModeFlags.Read);
            string internalText = file.GetAsText();
            file.Close();
            file.Open("res://jsonKeycodes/international-keycodes.json", Godot.File.ModeFlags.Read);
            string internationalText = file.GetAsText();
            file.Close();
            file.Open("res://jsonKeycodes/modifiers-keycodes.json", Godot.File.ModeFlags.Read);
            string modifiersText = file.GetAsText();
            file.Close();
            file.Open("res://jsonKeycodes/shifted-keycodes.json", Godot.File.ModeFlags.Read);
            string shiftedText = file.GetAsText();
            file.Close();
            file.Open("res://jsonKeycodes/layers-keycodes.json", Godot.File.ModeFlags.Read);
            string layersText = file.GetAsText();
            file.Close();
            file.Open("res://jsonKeycodes/led-keycodes.json", Godot.File.ModeFlags.Read);
            string ledText = file.GetAsText();
            file.Close();
            file.Open("res://jsonKeycodes/custom-keycodes.json", Godot.File.ModeFlags.Read);
            string customText = file.GetAsText();
            file.Close();
            Basic = JsonConvert.DeserializeObject<List<KeyCode>>(basictext).ToDictionary(keycode => keycode.Code);
            LessUsed = JsonConvert.DeserializeObject<List<KeyCode>>(lessUsedtext).ToDictionary(keycode => keycode.Code);
            Bluetooth = JsonConvert.DeserializeObject<List<KeyCode>>(bluetoothtext).ToDictionary(keycode => keycode.Code);
            InternalCodes = JsonConvert.DeserializeObject<List<KeyCode>>(internalText).ToDictionary(keycode => keycode.Code);
            International = JsonConvert.DeserializeObject<List<KeyCode>>(internationalText).ToDictionary(keycode => keycode.Code);
            Modifiers = JsonConvert.DeserializeObject<List<KeyCode>>(modifiersText).ToDictionary(keycode => keycode.Code);
            Shifted = JsonConvert.DeserializeObject<List<KeyCode>>(shiftedText).ToDictionary(keycode => keycode.Code);
            Layers = JsonConvert.DeserializeObject<List<KeyCode>>(layersText).ToDictionary(keycode => keycode.Code);
            Led = JsonConvert.DeserializeObject<List<KeyCode>>(ledText).ToDictionary(keycode => keycode.Code);

            CustomCodes = JsonConvert.DeserializeObject<List<KeyCode>>(customText).ToDictionary(keycode => keycode.Code);
            addBlankSubKeys();

        }
        public static KeyCodes Instance()
        {
            // Uses lazy initialization
            // Note: this is not thread safe.
            if (instance == null)
            {
                instance = new KeyCodes();
            }
            return instance;
        }
        void addBlankSubKeys()
        {
            KeyCode noKey = BaseKeyCodeForString("KC.NO");
            for (int i = 0; i < Modifiers.Values.Count; i++)    
            {
                KeyCode keycode = Modifiers.ElementAt(i).Value;
                if (keycode.CanHaveSub)
                {
                    keycode.SubOne = noKey;
                }

            }
        }
        // todo not dry clean up
        public KeyCode BaseKeyCodeForString(string code)
        {
            bool hasValue = Basic.TryGetValue(code, out KeyCode value);
            if (hasValue)
            {
                return value.NewKeyCode();
            }
            else
            {
                return null;
            }
        }
        public KeyCode LedKeyCodeForString(string code)
        {
            bool hasValue = Led.TryGetValue(code, out KeyCode value);
            if (hasValue)
            {
                return value.NewKeyCode();
            }
            else
            {
                return null;
            }
        }
        public KeyCode LessUsedKeyCodeForString(string code)
        {
            bool hasValue = LessUsed.TryGetValue(code, out KeyCode value);
            if (hasValue)
            {
                return value.NewKeyCode();
            }
            else
            {
                return null;
            }
        }
        public KeyCode BluetoothKeyCodeForString(string code)
        {
            bool hasValue = Bluetooth.TryGetValue(code, out KeyCode value);
            if (hasValue)
            {
                return value.NewKeyCode();
            }
            else
            {
                return null;
            }
        }
        public KeyCode InternalKeyCodeForString(string code)
        {
            bool hasValue = InternalCodes.TryGetValue(code, out KeyCode value);
            if (hasValue)
            {
                return value.NewKeyCode();
            }
            else
            {
                return null;
            }
        }
        public KeyCode LayerKeyCodeForString(string code)
        {
            bool hasValue = Layers.TryGetValue(code, out KeyCode value);
            if (hasValue)
            {
                return value.NewKeyCode();
            }
            else
            {
                return null;
            }
        }
        public KeyCode InternationalKeyCodeForString(string code)
        {
            bool hasValue = International.TryGetValue(code, out KeyCode value);
            if (hasValue)
            {
                return value.NewKeyCode();
            }
            else
            {
                return null;
            }
        }
        public KeyCode ModifiersKeyCodeForString(string code)
        {
            bool hasValue = Modifiers.TryGetValue(code, out KeyCode value);
            if (hasValue)
            {
                return value.NewKeyCode();
            }
            else 
            {
                if (code.Contains("("))
                {
                    var splitCode = code.Split("(");
                    string newSeachCode = splitCode[0] + "(kc)";
                    bool secondSearch = Modifiers.TryGetValue(newSeachCode, out KeyCode SecondValue);
                    if (secondSearch)
                    {
                        string remvedstuff= splitCode[1].Substring(0, splitCode[1].Length - 1);
                        KeyCode NeededKey = SecondValue.NewKeyCode();
                        NeededKey.SubOne = KeyCodeForString(remvedstuff);
                        GD.Print(NeededKey.SubOne.Code);
                        return NeededKey;
                    }

                }

                return null;
            }
        }
        public KeyCode ShiftedKeyCodeForString(string code)
        {
            bool hasValue = Shifted.TryGetValue(code, out KeyCode value);
            if (hasValue)
            {
                return value.NewKeyCode();
            }
            else
            {
                return null;
            }
        }
        public KeyCode CustomKeyCodeForString(string code, string dirtycode)
        {
            bool hasValue = CustomCodes.TryGetValue(code, out KeyCode value);
            if (hasValue)
            {
                return value.NewKeyCode();
            }
            else
            {
                bool hasValueSecond = CustomCodes.TryGetValue(dirtycode, out KeyCode valueSecond);
                if (hasValueSecond)
                {
                    return valueSecond.NewKeyCode();
                }
                else
                {
                    return null;

                }
            }
        }
        private static readonly Regex sWhitespace = new Regex(@"\s+");
        public static string ReplaceWhitespace(string input, string replacement)
        {
            return sWhitespace.Replace(input, replacement);
        }
        public KeyCode KeyCodeForString(string dirtycode)
        {
            string code = ReplaceWhitespace(dirtycode, "");
            KeyCode tempCode;
           tempCode= BaseKeyCodeForString(code);
            if (tempCode != null)
            {
                return tempCode;
            }
            tempCode = ShiftedKeyCodeForString(code);
            if (tempCode != null)
            {
                return tempCode;
            }
            tempCode = ModifiersKeyCodeForString(code);
            if (tempCode != null)
            {
                return tempCode;
            }
            tempCode = LessUsedKeyCodeForString(code);
            if (tempCode != null)
            {
                return tempCode;
            }
            tempCode = InternationalKeyCodeForString(code);
            if (tempCode != null)
            {
                return tempCode;
            }
            tempCode = InternalKeyCodeForString(code);
            if (tempCode != null)
            {
                return tempCode;
            }
            tempCode = LedKeyCodeForString(code);
            if (tempCode != null)
            {
                return tempCode;
            }
            tempCode = LayerKeyCodeForString(code);
            if (tempCode != null)
            {
                return tempCode;
            }
            tempCode = CustomKeyCodeForString(code, dirtycode);
            if (tempCode != null)
            {
                return tempCode;
            }
            else
            {
                KeyCode errorCode = new KeyCode();
                errorCode.Code = "KC.NO";
                errorCode.Display = "ERR ";
                errorCode.Keybinding = "";
                errorCode.Description = "error no key code '" + code+"' from imported map";
                errorCode.SubNumber = 0;
                errorCode.CanHaveSub = false;
                errorCode.CanHaveSubNumber = false;
                return errorCode;
                // return null;
            }
        }
        public void RemoveCustomCode(string code)
        {
            CustomCodes.Remove(code);
            this.saveCustomCodes();

        }
        public void AddCustomCode(KeyCode newCode)
        {
            CustomCodes.Add(newCode.Code, newCode);
            this.saveCustomCodes();
        }
        void saveCustomCodes()
        {
            Godot.File file = new Godot.File();
            file.Open("res://jsonKeycodes/custom-keycodes.json", Godot.File.ModeFlags.Write);
            string customText = JsonConvert.SerializeObject(new List<KeyCode>(this.CustomCodes.Values), Formatting.Indented);
            file.StoreString(customText);
            file.Close();
        }



    }
}

File: client/src/logic/keycodes.ts
import { ElectronEvents, KeyCode } from "../types/types";
import _basic from "../jsonKeycodes/basic-keycodes.json"
import _bluetooth from "../jsonKeycodes/bluetooth-keycodes.json"
import _custom from "../jsonKeycodes/custom-keycodes.json"
import _internal from "../jsonKeycodes/internal-keycodes.json"
import _international from "../jsonKeycodes/international-keycodes.json"
import _layers from "../jsonKeycodes/layers-keycodes.json"
import _led from "../jsonKeycodes/led-keycodes.json"
import _lessused from "../jsonKeycodes/lessused-keycodes.json"
import _modifiers from "../jsonKeycodes/modifiers-keycodes.json"
import _shifted from "../jsonKeycodes/shifted-keycodes.json"
import { ClientManager } from "./clientManager";

export class KeyCodes {
    private static instance: KeyCodes;
    basic: Map<string, KeyCode>;
    lessUsed: Map<string, KeyCode>;
    bluetooth: Map<string, KeyCode>;
    internalCodes: Map<string, KeyCode>;
    international: Map<string, KeyCode>;
    modifiers: Map<string, KeyCode>;
    shifted: Map<string, KeyCode>;
    layers: Map<string, KeyCode>;
    led: Map<string, KeyCode>;
    customCodes: Map<string, KeyCode>;

    private constructor() {

        //@ts-ignore
        const basicKeycodes = _basic as KeyCode[]
        this.basic = new Map(basicKeycodes.map(i => [i.code, i]));
        //@ts-ignore
        const bluetoothKeycodes = _bluetooth as KeyCode[]
        this.bluetooth = new Map(bluetoothKeycodes.map(i => [i.code, i]));
        //@ts-ignore
        const customKeycodes = _custom as KeyCode[]
        this.customCodes = new Map(customKeycodes.map(i => [i.code, i]));
        //@ts-ignore
        const internalKeycodes = _internal as KeyCode[]
        this.internalCodes = new Map(internalKeycodes.map(i => [i.code, i]));
        //@ts-ignore
        const internationalKeycodes = _international as KeyCode[]
        this.international = new Map(internationalKeycodes.map(i => [i.code, i]));
        //@ts-ignore
        const layersKeycodes = _layers as KeyCode[]
        this.layers = new Map(layersKeycodes.map(i => [i.code, i]));
        //@ts-ignore
        const ledKeycodes = _led as KeyCode[]
        this.led = new Map(ledKeycodes.map(i => [i.code, i]));
        //@ts-ignore
        const lessusedKeycodes = _lessused as KeyCode[]
        this.lessUsed = new Map(lessusedKeycodes.map(i => [i.code, i]));
        //@ts-ignore
        const modifiersKeycodes = _modifiers as KeyCode[]
        this.modifiers = new Map(modifiersKeycodes.map(i => [i.code, i]));
        //@ts-ignore
        const shiftedKeycodes = _shifted as KeyCode[]
        this.shifted = new Map(shiftedKeycodes.map(i => [i.code, i]));
        this.addBlankSubKeys();


    }

    public static getInstance(): KeyCodes {
        if (!KeyCodes.instance) {
            KeyCodes.instance = new KeyCodes();
        }
        return KeyCodes.instance;
    }
    addBlankSubKeys() {
        const noKey = this.CodeForStringAndSet("KC.NO", this.basic);
        this.modifiers.forEach(keycode => {
            if (keycode.canHaveSub) {
                keycode.subOne = noKey;
            }
        });
    }
    public CodeForStringAndSet(code: string, searchingSet: Map<string, KeyCode>): KeyCode | undefined {
        const tempCode = searchingSet.get(code)
        if (tempCode !== undefined) {
            return { ...tempCode }
        } else {
            return undefined
        }
    }
    customCodeForString(dirtyCode: string, code: string): KeyCode | undefined {
        let tempCode = this.customCodes.get(code)
        if (tempCode !== undefined) {

            return { ...tempCode }
        } else {

            tempCode = this.customCodes.get(dirtyCode)
            if (tempCode !== undefined) {
                return { ...tempCode }
            }
            return undefined
        }
    }

    ModifiersKeyCodeForString(code: string): KeyCode | undefined {
        const tempCode = this.modifiers.get(code);
        if (tempCode !== undefined) {
            return { ...tempCode }
        }
        else {
            if (code.includes("(")) {
                var splitCode = code.split("(");
                const newSeachCode = splitCode[0] + "(kc)";
                const secondSearch = this.modifiers.get(newSeachCode);
                if (secondSearch !== undefined) {
                    const remvedstuff = splitCode[1].substring(0, splitCode[1].length - 1);
                    const NeededKey = { ...secondSearch }
                    NeededKey.subOne = this.KeyCodeForString(remvedstuff);
                    return NeededKey;
                }
            }
            return undefined;
        }
    }

    public KeyCodeForString(dirtycode: string): KeyCode {
        const code: string = dirtycode.trim();

        let tempCode: KeyCode | undefined;
        tempCode = this.CodeForStringAndSet(code, this.basic);
        if (tempCode != undefined) {
            return tempCode;
        }
        tempCode = this.CodeForStringAndSet(code, this.shifted);
        if (tempCode != undefined) {
            return tempCode;
        }
        tempCode = this.ModifiersKeyCodeForString(code);
        if (tempCode != undefined) {
            return tempCode;
        }
        tempCode = this.CodeForStringAndSet(code, this.lessUsed);
        if (tempCode != undefined) {
            return tempCode;
        }
        tempCode = this.CodeForStringAndSet(code, this.international);
        if (tempCode != undefined) {
            return tempCode;
        }
        tempCode = this.CodeForStringAndSet(code, this.internalCodes);
        if (tempCode != undefined) {
            return tempCode;
        }
        tempCode = this.CodeForStringAndSet(code, this.layers);
        if (tempCode != undefined) {
            return tempCode;
        }
        tempCode = this.CodeForStringAndSet(code, this.led);
        if (tempCode != undefined) {
            return tempCode;
        }
        tempCode = this.customCodeForString(dirtycode, code);
        if (tempCode != undefined) {
            return tempCode;
        }
        else {
            const errorCode: KeyCode = {
                code: "KC.NO",
                display: "ERR",
                keybinding: "",
                canHaveSub: false,
                canHaveSubNumber: false,
                subNumber: 0,
                Description: "error no key code '" + code + "' from imported map"
            }
            console.log("returning a shit code")
            return errorCode;
        }
    }

    public RemoveCustomCode(code: string) {
        this.customCodes.delete(code);
        this.saveCustomCodes();

    }

    public AddCustomCode(newCode: KeyCode) {
        this.customCodes.set(newCode.code, newCode);
        this.saveCustomCodes();

    }

    saveCustomCodes() {
        const clientManager = ClientManager.getInstance();
        clientManager.sendToBackend(ElectronEvents.SaveCustomCodes,
            JSON.stringify(Array.from(this.customCodes.values())).replace("null", '""')
        )
    }
}